### PR TITLE
[Feature](Nereids) Simplify range of predicate

### LINKED ISSUE #14059: [Feature](Nereids) Simplify range of predicate
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

for example：
1. a > 1 or a > 2
=> a > 1

2. a in (1,2,3) or a (3,4,5)
=> a in (1,2,3,4,5)

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qzsee**
# Proposed changes

Issue Number: close #14059 

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [x] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.07 seconds
 load time: 444 seconds
 storage size: 17179719700 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221109100118_clickbench_pr_42674.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * This class implements the function to simplify expression range.
+ * for example:
+ * a > 1 and a > 2 => a > 2
+ * a > 1 or a > 2 => a > 1
+ * a in (1,2,3) and a > 1 => a in (1,2,3)
+ * a in (1,2,3) and a in (3,4,5) => a = 3
+ * a in(1,2,3) and a in (4,5,6) => a in(1,2,3) and a in (4,5,6)
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java:56]: please check all comments

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/SimplifyRangeTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+public class SimplifyRangeTest {
+
+    private static final NereidsParser PARSER = new NereidsParser();
+    private ExpressionRuleExecutor executor;
+
+    @Test
+    public void testSimplify() {
+        executor = new ExpressionRuleExecutor(ImmutableList.of(SimplifyRange.INSTANCE));
+        assertRewrite("TA", "TA");
+        assertRewrite("TA = 1 and TA > 10", "TA = 1 and TA > 10");
```
[the remark this answers is not in the document]
Comment (qzsee) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/SimplifyRangeTest.java:52]: TA = 1 and TA > 10 ==> false

[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/SimplifyRangeTest.java]: my fault

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+     * use `Set` to wrap `InPredicate`
+     * for example:
+     * a in (1,2,3) => [1,2,3]
+     */
+    private static class DiscreteValue extends ValueDesc {
+        Set<Literal> values;
+
+        public DiscreteValue(Expression reference, Expression expr, Set<Literal> values) {
+            super(reference, expr);
+            this.values = Sets.newLinkedHashSet(values);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java:330]: I suggest use `Sets.newTreeSet` in this class, to keep the order.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                if (value.equals(ValueDesc.EMPTY)) {
+                    result.add(predicate);
+                } else if (value instanceof ResultValue) {
+                    // With simplified expression, it is possible to build `ValueDesc` as well
+                    ValueDesc o = value.expr.accept(this, null);
+                    if (o instanceof ResultValue) {
+                        result.add(o.expr);
+                    } else {
+                        valueDescList.add(o);
+                    }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java:157]: Why simplify `resultValue.expr` again?
I think you can direct add to the result list: 
```java
else if (value instanceof ResultValue) {
  result.add(o.expr);
}
```

**Review by 924060929 — COMMENTED** [review]
The `ResultValue` and the `ValueDesc.EMPTY` has the same UNKNOWN properties, The orthogonality is not satisfied.
I suggest rename the `ResultValue` to `UnknownValue`, and remove  `ValueDesc.EMPTY`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+                    if (range.isConnected(o.range)) {
+                        RangeValue rangeValue = new RangeValue(reference, ExpressionUtils.or(expr, other.expr));
+                        rangeValue.range = range.span(o.range);
+                        return rangeValue;
+                    }
+                    return EMPTY;
+                }
+                return union(this, (DiscreteValue) other);
+            } catch (Exception e) {
+                return EMPTY;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyRange.java:274]: What exception will be throw? Please add example comment to explain

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/SimplifyRangeTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        assertRewrite("TA > 3 and TA < 1", "FALSE");
+        assertRewrite("TA >= 3 and TA < 3", "TA >= 3 and TA < 3");
+        assertRewrite("TA = 1 and TA > 10", "FALSE");
+        assertRewrite("TA > 5 or TA < 1", "TA > 5 or TA < 1");
+        assertRewrite("TA > 5 or TA > 1 or TA > 10", "TA > 1");
+        assertRewrite("TA > 5 or TA > 1 or TA < 10", "TA > 5 or TA > 1 or TA < 10");
+        assertRewrite("TA > 5 and TA > 1 and TA > 10", "TA > 10");
+        assertRewrite("TA > 5 and TA > 1 and TA < 10", "TA > 5 and TA < 10");
+        assertRewrite("TA > 1 or TA < 1", "TA > 1 or TA < 1");
+        assertRewrite("TA > 1 or TA < 10", "TA > 1 or TA < 10");
```
Comment (924060929) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/SimplifyRangeTest.java:63]: `TA > 1 or TA < 10` should simplified to `true`

**Review by 924060929 — APPROVED** [review]

**Comment by 924060929** [comment]
We need support multi range and other type(like string) later

**Pull request merged**